plugins/woocommerce/src/Internal/Caches/VersionStringsGenerator.php

```diff
@@ -0,0 +1,183 @@
+	 *
+	 * @param LegacyProxy $legacy_proxy Legacy proxy instance.
+	 */
+	final public function init( LegacyProxy $legacy_proxy ) {
+		$this->legacy_proxy = $legacy_proxy;
+	}
+
+	/**
+	 * Tells whether the version strings cache should be used or not.
+	 *
+	 * This will return true only if an external object cache is configured in WordPress,
+	 * since otherwise the cached entries will only persist for the current request.
+	 *
+	 * @return bool
+	 */
+	public function should_use(): bool {
+		if ( ! is_null( $this->should_use ) ) {
+			return $this->should_use;
+		}
+
+		$this->should_use = is_null( $this->legacy_proxy ) ? false : ( $this->legacy_proxy->call_function( 'wp_using_ext_object_cache' ) ?? false );
+
+		return $this->should_use;
+	}
+
+	/**
+	 * Get the current version string for an ID.
+	 *
+	 * If no version exists and $generate is true, a new version will be created.
+	 * If no version exists and $generate is false, null will be returned.
+	 *
+	 * @param string $id       The ID to get the version string for.
+	 * @param bool   $generate Whether to generate a new version if one doesn't exist. Default true.
+	 * @return string|null Version string, or null if not found and $generate is false.
+	 * @throws \InvalidArgumentException If id is invalid.
+	 *
+	 * @since 10.4.0
+	 */
+	public function get_version( string $id, bool $generate = true ): ?string {
+		$this->validate_input( $id );
+
+		$cache_key = $this->get_cache_key( $id );
+		$found     = false;
+		$version   = wp_cache_get( $cache_key, self::CACHE_GROUP, false, $found );
+
+		if ( ! $found ) {
+			if ( ! $generate ) {
+				return null;
+			}
+			$version = $this->generate_version( $id );
+		} else {
+			// Refresh the cache lifetime.
+			$this->store_version( $id, $version );
+		}
+		return $version;
+	}
+
+	/**
+	 * Generate and store a new version string for an ID.
+	 * The already existing version string, if any, will be replaced.
+	 *
+	 * @param string $id The ID to generate a version string for.
+	 * @return string The new version string.
+	 * @throws \InvalidArgumentException If id is invalid.
+	 *
+	 * @since 10.4.0
+	 */
+	public function generate_version( string $id ): string {
+		$this->validate_input( $id );
+
+		$version = wp_generate_uuid4();
+		$this->store_version( $id, $version );
+		return $version;
+	}
+
+	/**
+	 * Store the version string in cache with a filterable TTL.
+	 *
+	 * @param string $id      The ID to store the version string for.
+	 * @param string $version The version string to store.
+	 * @return bool True on success, false on failure.
+	 */
+	protected function store_version( string $id, string $version ): bool {
+		$cache_key = $this->get_cache_key( $id );
+
+		/**
+		 * Filter the TTL for version string cache.
+		 *
+		 * @param int    $ttl Time to live in seconds. Default 1 day.
+		 * @param string $id  The ID.
+		 *
+		 * @since 10.4.0
+		 */
+		$ttl = apply_filters( 'woocommerce_version_string_ttl', DAY_IN_SECONDS, $id );
```

Comment:
Very minor but this name might get confused with the actual "woocommerce_version."

Should we rename to:

```suggestion
		$ttl = apply_filters( 'woocommerce_version_string_generator_ttl', DAY_IN_SECONDS, $id );
```